### PR TITLE
sh_string was broken in some cases, adds functionality to fix it. Fixes #455

### LINKED ISSUE #455: `pwnlib.misc.util.sh_string` broken for newline
If you pass *only* a newline to `sh_string`, the result is incorrect.  This may affect other input.

```py
>>> with open('\n','w+') as f: f.write("HELLO\n")
>>> exitcode = os.system("cat '\n'")
HELLO
>>> exitcode = os.system("cat %s" % sh_string('\n'))
cat: : No such file or directory
```

### PR DESCRIPTION
**Pull request by Idolf**
This should hopefully fix #455.

### CODE REVIEW

**Comment by zachriggle** [comment]
It's a bummer that BSD sh doesn't support arrays :(

```
declare -a argv
argv[0]=echo
argv[1]=hello
argv[2]=world
exec "${argv[@]}"
```

File: pwnlib/util/misc.py [hunk 1]
```diff
@@ -270,6 +271,69 @@ def mkdir_p(path):
...
+    very_good = set(string.ascii_letters + string.digits)
+    good      = (very_good | set(string.punctuation + ' ')) - set("'")
+    alt_good  = (very_good | set(string.punctuation + ' ')) - set('!')
+
+    if '\x00' in s:
+        log.error("sh_string(): Cannot create a null-byte")
+
+    export = 'export ' if export else ''
+
+    if all(c in very_good for c in s):
```
Comment (zachriggle) [inline pwnlib/util/misc.py]: I feel like we should be able to skip almost all of this logic and just use sh_string for both.  This is basically duplicating the same logic.

sh-3.2$ export $(echo hello)=$(echo world)
sh-3.2$ echo $hello
world

**Comment by Idolf** [comment]
It still would not fix the core issue, which is that it is very hard to create a newline-terminated string using only posix features. Strings inside arrays also have this problem.

File: pwnlib/util/misc.py [hunk 2]
```diff
@@ -270,6 +271,69 @@ def mkdir_p(path):
         else:
             raise
 
+def sh_prepare(var, s, export = False):
```
Comment (zachriggle) [inline pwnlib/util/misc.py]: Wouldn't it make more sense for this to just take a dictionary?

**Comment by zachriggle** [comment]
True re:
https://github.com/Gallopsled/pwntools/pull/667#issuecomment-240827612

**Comment by zachriggle** [comment]
Is there any way to compile the various implementations of `sh` for 32-bit Linux and just carry them around in a test suite?

**Comment by Idolf** [comment]
No clue with regards to the test suite. However it only uses POSIX features, which should hopefully be implemented correctly in all shells.

File: pwnlib/util/misc.py (hunk 2 above)
Comment (Idolf) [inline pwnlib/util/misc.py]: Sure. I can do a new commit based on this.

File: pwnlib/util/misc.py (hunk 1 above)
Comment (Idolf) [inline pwnlib/util/misc.py]: Done.

File: pwnlib/util/misc.py (hunk 2 above)
Comment (Idolf) [inline pwnlib/util/misc.py]: Done.

**Comment by Idolf** [comment]
This is ready to merge.

**Comment by zachriggle** [comment]
On mobile, merge away

**Pull request merged**